### ISSUE TITLE
Compilation issue with gcc@12.1.0

### ISSUE THREAD

**Issue by wdconinc** [KATRIN-Experiment/Kassiopeia#60]
Compiling Kassiopeia with gcc@12.1.0 and c++17 fails (possibly with `-Werror -Wall`) due to the deprecation of the `std::iterator` types in c++17. This is a fairly straighforward fix, and I can take a stab at it. 

```
In file included from /home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSortedSurfaceContainer.hh:4,
                 from /home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/src/KSortedSurfaceContainer.cc:1:
/home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurfaceContainer.hh:55:34: error: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Werror=deprecated-declarations]
   55 |     class iterator : public std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>
      |                                  ^~~~~~~~
In file included from /usr/include/c++/12/string:45,
                 from /home/wdconinc/git/Kassiopeia/KEMField/Source/Core/include/KFundamentalTypes.hh:6,
                 from /home/wdconinc/git/Kassiopeia/KEMField/Source/Core/include/KDataComparator.hh:4,
                 from /home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurfaceID.hh:5,
                 from /home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurface.hh:4,
                 from /home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurfaceContainer.hh:4:
/usr/include/c++/12/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
/home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurfaceContainer.hh:59:33: error: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Werror=deprecated-declarations]
   59 |         using value_type = std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>::value_type;
      |                                 ^~~~~~~~
/usr/include/c++/12/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
/home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurfaceContainer.hh:60:32: error: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Werror=deprecated-declarations]
   60 |         using reference = std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>::reference;
      |                                ^~~~~~~~
/usr/include/c++/12/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
/home/wdconinc/git/Kassiopeia/KEMField/Source/Surfaces/include/KSurfaceContainer.hh:61:30: error: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Werror=deprecated-declarations]
   61 |         using pointer = std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>::pointer;
      |                              ^~~~~~~~
/usr/include/c++/12/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
```

**Comment by wdconinc** [comment]
Something like this diff:
```diff
diff --git a/KEMField/Source/Surfaces/include/KSurfaceContainer.hh b/KEMField/Source/Surfaces/include/KSurfaceContainer.hh
index 21eecca2..b15f792a 100644
--- a/KEMField/Source/Surfaces/include/KSurfaceContainer.hh
+++ b/KEMField/Source/Surfaces/include/KSurfaceContainer.hh
@@ -52,13 +52,24 @@ class KSurfaceContainer
 * @author T.J. Corona
 */
 
+#if __cplusplus >= 201700L
+    class iterator
+#else
     class iterator : public std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>
+#endif
     {
       public:
+#if __cplusplus >= 201700L
+        using iterator_category = std::bidirectional_iterator_tag;
+        using value_type = KSurfacePrimitive*;
+        using reference = KSurfacePrimitive*&;
+        using pointer = KSurfacePrimitive**;
+#else
         using iterator_category = std::bidirectional_iterator_tag;
         using value_type = std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>::value_type;
         using reference = std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>::reference;
         using pointer = std::iterator<std::bidirectional_iterator_tag, KSurfacePrimitive*>::pointer;
+#endif
 
         friend class KSurfaceContainer;

```
possibly with a better c++17 detection, or going back to whenever the member type interface was introduced.

**Comment by wdconinc** [comment]
Probably `#if __cplusplus >= 201703L` is more correct.

**Comment by zykure** [comment]
Thanks for the report and fix!

The `testing` branch should not have this issue because I made a similar fix there when I encountered this issue earlier. However I like your implementation better, because it keeps backwards compatibility to C++11. I will merge #62 after some tests.

(The `testing` branch receives updates more regularly than `main`. My intention with this was to have a faster update cycle when we change code in the KATRIN repositories, since `main` is tied to new releases which only happen every so often. `testing` on the other hand is mostly in sync with KATRIN development.)
